File: 127-word-ladder/word-ladder.cpp
class Solution {
public:
    int diff(string s,string p)
    {
        int c=0;
        for(int i=0;i<s.size();i++)
        {
            if(s[i]!=p[i])
            {
                c++;
            }
        }
    return c;
    }
    // long fun(string beginWord, string endWord, vector<string>& wordList,int i)
    // {
    //     if(beginWord==endWord)
    //     {
    //         return 0;
    //     }
    //     if(i>=wordList.size())
    //     {
    //         return INT_MAX;
    //     }
    //     long ans=INT_MAX;
    //     for(int j=i;j<wordList.size();j++)
    //     {
    //         if(diff(wordList[j],beginWord)==1)
    //         {
    //             ans=min(ans, 1 + fun(wordList[j],endWord,wordList,j+1));

    //         }
    //     }
    //     return ans;
    // }
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        unordered_map<string,int> m;
         for(int i=0;i<wordList.size();i++)
            {
                m[wordList[i]]=INT_MAX;
            }
        queue<string> q;
        long ans=INT_MAX;
        m[beginWord]=0;
        q.push(beginWord);
        while(!q.empty())
        {
            string p=q.front();
            long cost=m[p];
            q.pop();
            if(p==endWord)
            {
                ans=min(ans,cost);
            }
            for(int i=0;i<wordList.size();i++)
            {
                if(diff(wordList[i],p)==1 && cost+1<m[wordList[i]])
                {
                    m[wordList[i]]=1+cost;
                    q.push(wordList[i]);
                }
            }
        }
        if(ans==INT_MAX)
        {
            return 0;
        }
        return ans+1;
    }
};

File: 127-word-ladder/word-ladder.java
class Solution {
    private boolean check(String a, String b) {
        int c = 0;
        for (int i = 0; i < a.length(); i++) {
            if (a.charAt(i) != b.charAt(i)) {
                c++;
                if (c > 1) {
                    return false;
                }
            }
        }
        if (c == 1)
            return true;

        return false;
    }

    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        wordList.add(0, beginWord);
        int n = wordList.size();
        List<List<Integer>> ls = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            ls.add(new ArrayList<>());
        }
        for (int i = 0; i < n; i++) {

            for (int j = i + 1; j < n; j++) {
                if (check(wordList.get(i), wordList.get(j))) {
                    ls.get(i).add(j);
                    ls.get(j).add(i);
                }
            }

        }
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[0] - b[0]);
        boolean[] visited = new boolean[n];
        int ans = 0;
        pq.offer(new int[] { 1, 0 });
        while (!pq.isEmpty()) {
            int[] top = pq.poll();
            int cost = top[0];
            int node = top[1];
            if (visited[node])
                continue;
            visited[node] = true;
            for (int i : ls.get(node)) {
                if (wordList.get(i).equals(endWord)) {
                    return cost + 1;
                }
                pq.offer(new int[] { cost + 1, i });
            }
        }
        return 0;
    }
}